File: 33.go
package main

import (
	"fmt"
	_ "fmt"
)

func binartSearch(nums []int, target int) int {
	if nums == nil || len(nums) == 0 || (len(nums) == 1 && target != nums[0]) {
		return -1
	}

	mid := len(nums) / 2
	if target < nums[mid] {
		return binartSearch(nums[:mid], target)
	} else if target > nums[mid] {
		ret := binartSearch(nums[mid+1:], target)
		if ret != -1 {
			ret += mid + 1
		}
		return ret
	} else {
		return mid
	}
}

func search(nums []int, target int) int {
	/* 选找中断点 */
	lens := len(nums)
	var pos int
	ret := 0
	i := 1
	for ; i < lens; i++ {
		if nums[i] < nums[i-1] {
			break
		}
	}
	if i == lens {
		pos = -1 /* 表示已经有序 */
	} else {
		pos = i /* 表示 [i,lens) 后有序，[0,i) 前有序 */
	}
	/* 二分查找 */
	if pos != -1 {
		if target > nums[lens-1] {
			ret = binartSearch(nums[:pos], target)
		} else if target < nums[lens-1] {
			ret = binartSearch(nums[pos:], target)
			if ret != -1 {
				ret += pos
			}
		} else {
			ret = lens - 1
		}
	} else {
		ret = binartSearch(nums, target)
	}
	return ret
}

func main() {
	nums := []int{1, 3, 5}
	target := 5
	ret := search(nums, target)
	fmt.Print(ret)
}

File: 33.cpp
#include <bits/stdc++.h>
#include <gtest/gtest.h>

using namespace std;
/* g++ x.cpp -lgtest -lgtest_main */
class Solution {
public:
  int search(vector<int> &nums, int target) {
    int lens = nums.size();
    int b = 0, e = lens - 1;
    while (b <= e) {
      int mid = (b + e) >> 1;
      if (nums[mid] >= nums[0]) {
        if (nums[0] <= target && target <= nums[mid]) {
          e = mid - 1;
        } else {
          b = mid + 1;
        }
      } else if (nums[mid] <= nums[lens - 1]) {
        if (nums[lens - 1] >= target && target >= nums[mid]) {
          b = mid + 1;
        } else {
          e = mid - 1;
        }
      }
    }
    if (b >= 0 && b <= lens - 1 && nums[b] == target)
      return b;
    if (e >= 0 && e <= lens - 1 && nums[e] == target)
      return e;
    return -1;
  }
};

TEST(Solution, isMatch) {
  Solution s;
  vector<int> v{4, 5, 6, 7, 0, 1};
  EXPECT_EQ(s.search(v, 0), 4);
  EXPECT_EQ(s.search(v, 3), -1);
  vector<int> v2{1};
  EXPECT_EQ(s.search(v2, 0), -1);
  vector<int> v3{1,3};
  EXPECT_EQ(s.search(v3, 1), 0);
}
